### ISSUE TITLE
TypeError: TypeMap.__get_cls_dict.<locals>.__init__: missing argument 'help'

### ISSUE THREAD

**Issue by yarikoptic** [hdmf-dev/hdmf#541]
I think I have not seen this one before and found no issue in hdmf-dev or `NeurodataWithoutBorders/` orgs

Upon reading of nwb file from http://github.com/dandisets/000019 leading to

```shell
drogon:/mnt/backup/dandi/dandisets/000019
$> dandi validate sub-EC2/sub-EC2_ses-EC2-B1.nwb
/home/yoh/miniconda3/envs/dandi-devel/lib/python3.8/site-packages/hdmf/spec/namespace.py:484: UserWarning: Ignoring cached namespace 'core' version 2.0.2 because version 2.2.5 is already loaded.
  warn("Ignoring cached namespace '%s' version %s because version %s is already loaded."
2021-03-18 09:46:44,029 [   ERROR] sub-EC2/sub-EC2_ses-EC2-B1.nwb: 7 error(s)
2021-03-18 09:46:44,029 [   ERROR]   VectorIndex/description (intervals/invalid_times/tags_index): argument missing
2021-03-18 09:46:44,029 [   ERROR]   VectorIndex (intervals/invalid_times/tags_index): incorrect type - expected 'uint8', got 'int64'
2021-03-18 09:46:44,029 [   ERROR]   VectorIndex/description (intervals/invalid_times/timeseries_index): argument missing
2021-03-18 09:46:44,029 [   ERROR]   VectorIndex (intervals/invalid_times/timeseries_index): incorrect type - expected 'uint8', got 'int64'
2021-03-18 09:46:44,029 [   ERROR]   VectorIndex/description (intervals/invalid_times/timeseries_index): argument missing
2021-03-18 09:46:44,029 [   ERROR]   VectorIndex (intervals/invalid_times/timeseries_index): incorrect type - expected 'uint8', got 'int64'
2021-03-18 09:46:44,029 [   ERROR]   Failed to read metadata: (root/processing/ecephys/spectrum GroupBuilder {'attributes': {'namespace': 'ndx-spectrum', 'neurodata_type': 'Spectrum'}, 'groups': {}, 'datasets': {'frequencies': root/processing/ecephys/spectrum/frequencies DatasetBuilder {'attributes': {}, 'data': <Closed HDF5 dataset>}, 'power': root/processing/ecephys/spectrum/power DatasetBuilder {'attributes': {}, 'data': <Closed HDF5 dataset>}}, 'links': {'source_timeseries': root/processing/ecephys/spectrum/source_timeseries LinkBuilder {'builder': root/acquisition/ElectricalSeries GroupBuilder {'attributes': {'comments': 'no comments', 'description': 'all Wav data', 'help': 'Stores acquired voltage data from extracellular recordings', 'namespace': 'core', 'neurodata_type': 'ElectricalSeries'}, 'groups': {}, 'datasets': {'data': root/acquisition/ElectricalSeries/data DatasetBuilder {'attributes': {'conversion': 0.001, 'resolution': 0.0, 'unit': 'volt'}, 'data': <Closed HDF5 dataset>}, 'electrodes': root/acquisition/ElectricalSeries/electrodes DatasetBuilder {'attributes': {'description': 'ECoG electrodes on brain', 'help': 'a subset (i.e. slice or region) of a DynamicTable', 'namespace': 'core', 'neurodata_type': 'DynamicTableRegion', 'table': root/general/extracellular_ephys/electrodes GroupBuilder {'attributes': {'colnames': array([b'x', b'y', b'z', b'imp', b'location', b'filtering', b'group',
       b'group_name', b'bad'], dtype='|S13'), 'description': 'metadata about extracellular electrodes', 'help': 'A column-centric table', 'namespace': 'core', 'neurodata_type': 'DynamicTable'}, 'groups': {}, 'datasets': {'bad': root/general/extracellular_ephys/electrodes/bad DatasetBuilder {'attributes': {'description': 'electrode identified as too noisy', 'help': 'Values for a list of elements', 'namespace': 'core', 'neurodata_type': 'VectorData'}, 'data': <Closed HDF5 dataset>}, 'filtering': root/general/extracellular_ephys/electrodes/filtering DatasetBuilder {'attributes': {'description': 'description of hardware filtering', 'help': 'Values for a list of elements', 'namespace': 'core', 'neurodata_type': 'VectorData'}, 'data': <Closed HDF5 dataset>}, 'group': root/general/extracellular_ephys/electrodes/group DatasetBuilder {'attributes': {'description': 'a reference to the ElectrodeGroup this electrode is a part of', 'help': 'Values for a list of elements', 'namespace': 'core', 'neurodata_type': 'VectorData'}, 'data': <hdmf.backends.hdf5.h5_utils.BuilderH5ReferenceDataset object at 0x7f964dcc0970>}, 'group_name': root/general/extracellular_ephys/electrodes/group_name DatasetBuilder {'attributes': {'description': 'the name of the ElectrodeGroup this electrode is a part of', 'help': 'Values for a list of elements', 'namespace': 'core', 'neurodata_type': 'VectorData'}, 'data': <Closed HDF5 dataset>}, 'id': root/general/extracellular_ephys/electrodes/id DatasetBuilder {'attributes': {'help': 'unique identifiers for a list of elements', 'namespace': 'core', 'neurodata_type': 'ElementIdentifiers'}, 'data': <Closed HDF5 dataset>}, 'imp': root/general/extracellular_ephys/electrodes/imp DatasetBuilder {'attributes': {'description': 'the impedance of the channel', 'help': 'Values for a list of elements', 'namespace': 'core', 'neurodata_type': 'VectorData'}, 'data': <Closed HDF5 dataset>}, 'location': root/general/extracellular_ephys/electrodes/location DatasetBuilder {'attributes': {'description': 'the location of channel within the subject e.g. brain region', 'help': 'Values for a list of elements', 'namespace': 'core', 'neurodata_type': 'VectorData'}, 'data': <Closed HDF5 dataset>}, 'x': root/general/extracellular_ephys/electrodes/x DatasetBuilder {'attributes': {'description': 'the x coordinate of the channel location', 'help': 'Values for a list of elements', 'namespace': 'core', 'neurodata_type': 'VectorData'}, 'data': <Closed HDF5 dataset>}, 'y': root/general/extracellular_ephys/electrodes/y DatasetBuilder {'attributes': {'description': 'the y coordinate of the channel location', 'help': 'Values for a list of elements', 'namespace': 'core', 'neurodata_type': 'VectorData'}, 'data': <Closed HDF5 dataset>}, 'z': root/general/extracellular_ephys/electrodes/z DatasetBuilder {'attributes': {'description': 'the z coordinate of the channel location', 'help': 'Values for a list of elements', 'namespace': 'core', 'neurodata_type': 'VectorData'}, 'data': <Closed HDF5 dataset>}}, 'links': {}}}, 'data': <Closed HDF5 dataset>}, 'starting_time': root/acquisition/ElectricalSeries/starting_time DatasetBuilder {'attributes': {'rate': 3051.7578, 'unit': 'Seconds'}, 'data': 0.0}}, 'links': {}}}}}, "Could not construct Spectrum object due to: TypeMap.__get_cls_dict.<locals>.__init__: missing argument 'help'")
Summary: Validation errors in 1 out of 1 files
2021-03-18 09:46:44,030 [    INFO] Logs saved in /home/yoh/.cache/dandi-cli/log/20210318134639Z-15851.log
```
or better printed on https://github.com/dandi/dandi-api-datasets/blob/32266dc5ad052d71a4ea55d1eee8b67bfbdcb3d5/000019/CONVERSION.errors

similar errors in other files/dandisets:

```shell
(git-annex)lena:~/proj/dandi/dandi-api-datasets[master]git-annex
$> git grep -l 'missing.*help'
000012/CONVERSION.errors
000019/CONVERSION.errors
```

DueCredit Report:
- Client to interact with DANDI Archive / dandi-cli (v 0.11.0+24.gb813347) [1]
- HDMF: Hierarchical Data Modeling Framework for Modern Science Data Standards / hdmf/ (v 2.3.0) [2]
- Scientific tools library / numpy (v 1.18.1) [3]

and pynwb 1.4.0

**Comment by yarikoptic** [comment]
oh, found this in https://github.com/NeurodataWithoutBorders/pynwb/issues/1341#issuecomment-788180764

**Comment by satra** [comment]
this is an important issue for us as this has made a rather important dataset unopenable. 

> curl -o trythis.nwb https://dandiarchive.s3.amazonaws.com/girder-assetstore/a4/36/a43656b9d3f9453aaa1e653c4eeb0226

a fix to this issue would be much appreciated, if you want more files to check you can download the entire dataset with `dandi download https://dandiarchive.org/dandiset/000008`

**Comment by rly** [comment]
I can confirm that this error exists for HDMF 2.4.0. Also, validation with HDMF 2.4.0 results in `RecursionError: maximum recursion depth exceeded while calling a Python object`.

I believe this is fixed in the latest dev branch of hdmf, due to changes in https://github.com/hdmf-dev/hdmf/pull/535. Please test if you can. I tested this on dandisets 12 and 19 and no longer get the above errors. (There are other validation errors though.)

**Comment by rly** [comment]
For the sake of record-keeping, the error was due to `TypeMap.get_container_cls(...)` not removing the `help` attribute from the docval for the generated `__init__` method. It does properly prevent the `help` attribute from being added as a field for the new class. The new code in #535 fixes this by excluding any container fields mapped to the name `help` in a single place, earlier in class generation. This could be done even earlier, at the level of the spec, and should be documented as well. Alternatively, we could remove this special case and allow `help` as a field. It does not seem necessary to keep this special case.

**Comment by rly** [comment]
@oruebel @ajtritt Do you remember why container fields named `help` were specially excluded for custom class generation?

**Comment by rly** [comment]
NVM I found it in #25. Also this bug was introduced in HDMF 2.3.0 from the refactoring in #420.

**Comment by rly** [comment]
This is fixed in the latest release HDMF 2.5 and PyNWB 1.5
